src/wasm-type.h

```diff
@@ -672,6 +672,11 @@ struct TypeBuilder {
   // relationships are not valid.
   BuildResult build();
 
+  // Like `build`, but first automatically groups types into rec groups and
+  // sorts the groups to create a valid type ordering, if possible. Explicitly
+  // set up rec groups will not be split up, but they may be merged.
+  BuildResult groupAndBuild();
```

Comment:
How does this automatic grouping work? I'm assuming that's only relevant if the user never called `createRecGroup` - if not, how does that interact?

Reply:
That's what Tarjan's algorithm is for. The strongly connected components in the type definition graph are turned into rec groups. Explicitly defined rec groups are represented using synthetic edges in the input to Tarjan's to make sure all the types in an input rec group are in the same strongly connected component and therefore in the same output rec group.

Reply:
I guess I just don't understand how this works. The existing types are all in rec groups, correct? (Looks like the PR changes to in `type-updating.cpp` assume that, in fact.) It seems odd to me that there is a way to infer grouping for new types added after them - that's a new problem that didn't exist before, and it's not obvious to me how it is solved. I get (sort of :laughing: ) what Tarjan does, but it makes the assumption that it *can* find a grouping without the user providing one. Why is that?

Reply:
(IOW, what I'm trying to ask is - why is the result here guaranteed to be what the user intended? Since the user did not provide more information, the problem seems too ambiguous to me. I'm definitely not looking at this right...)